lib: expose default prepareStackTrace (#50827)
nodejs/node: doc/api/module.md

`path` is the resolved path for the file for which a corresponding source map
should be fetched.

> ### `module.prepareStackTrace(error, trace)`

Reviewer: Why is this in `node:module` though? I'd expect it to be in `util`, as it can be used in other use cases beyond modules?

Author: We have other source maps related methods exposed in `node:module`, like `module.findSourceMap` and `module.SourceMap`. It's my first impression that this method can be in `node:module`.

Reviewer: This seems to be more about errors in general, and not limited to source maps, we also have `util.getSystemErrorName(err)` etc. in util. I can imagine users wanting to use it for other purposes (e.g. add a hook to get the structured trace without removing the default stack formatting once and for all - isn't this actually what our internal `overrideStackTrace` thing do? I can actually see us replacing `overrideStackTrace` with this internally).

Reviewer: Actually we probably cannot replace `overrideStackTrace` because we still care about other people's callback. Speaking of which, can't we just expose this as the default value of `Error.prepareStackTrace`? This is ignored by V8 because we use the C++ API, so we can hang our default implementation there. Then users just do

```js
Error.prepareStackTrace = (error, trace) => {
  const defaultCallback = Error.prepareStackTrace;
  if (defaultCallback !== undefined) {
    // do what they do...
  }
  return defaultCallback(error, trace);
}
```

Author: Yes, it is true that exposing the default as `Error.prepareStackTrace` can be more intuitive. I've fixed up incorrect stack trace array modification in REPL and updated the PR.

<!-- YAML
added: REPLACEME
-->

* `error` {Object}
* `trace` {Object\[]}
* Returns: {string} Returns formatted stack trace.

This is the default implementation Node.js stack trace formatter. When source

> map support is enabled, this function will attempt to map the stack trace to

Reviewer: We should update the section in `cli.md` https://github.com/nodejs/node/blob/02db7fc2cd3cdfc2b097a983449b33f54735df4b/doc/api/cli.md#L526-L527

the original source locations. This can be used to override the default stack
trace formatter.

It is discouraged to modify the stack trace as it can lead to confusion when
debugging with the stack traces.

```mjs

> import { prepareStackTrace } from 'node:module';
> Error.prepareStackTrace = (error, trace) => {
>   // Filter out custom entries.
>   trace = trace.filter((it) => {
>     return !it.getFileName().includes('my-internal-module');
>   });
>   return prepareStackTrace(error, trace);
> };

Reviewer: I think we should be cautious about documenting the API here, because the stability of these arguments are not guaranteed by us. A link to https://v8.dev/docs/stack-trace-api should be enough, and we can just put it down as

```suggestion
import { prepareStackTrace: defaultPrepareStackTrace } from 'node:module';
Error.prepareStackTrace = (error, trace) => {
  // Modify error and trace.
  return defaultPrepareStackTrace(error, trace);
};
```

(Or maybe just export this as `defaultPrepareStackTrace`)
